src/client/serialize/daos_serialize.c

```diff
@@ -0,0 +1,1064 @@
+	if (close_cont)
+		daos_cont_close(coh, NULL);
+	daos_prop_free(label);
+	return rc;
+}
+
+int
+daos_cont_deserialize_props(daos_handle_t poh, char *filename,
+			    daos_prop_t **props, uint64_t *cont_type)
+{
+	int	rc = 0;
+	hid_t	file_id = 0;
+
+	D_PRINT("Reading metadata file: %s\n", filename);
+
+	file_id = H5Fopen(filename, H5F_ACC_RDONLY, H5P_DEFAULT);
+	if (file_id < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to open metadata file: %s\n", filename);
+		D_GOTO(out, rc);
+	}
+	rc = deserialize_props(poh, file_id, props, cont_type);
+	if (rc != 0) {
+		D_ERROR("failed to deserialize cont props "DF_RC"\n",
+			DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+out:
+	if (file_id > 0)
+		H5Fclose(file_id);
+	return rc;
+}
+
+static int
+deserialize_attrs(hid_t file_id, uint64_t *_num_attrs,
+		  char ***_names, void ***_buffers, size_t **_sizes)
+{
+	hid_t		status = 0;
+	int		rc = 0;
+	int		num_attrs = 0;
+	int		num_dims = 0;
+	char		**names = NULL;
+	void		**buffers = NULL;
+	size_t		*sizes = NULL;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hid_t		vtype = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	/* Read the user attributes */
+	dset = H5Dopen(file_id, "User Attributes", H5P_DEFAULT);
+	if (dset < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to open User Attributes Dataset\n");
+		D_GOTO(out, rc);
+	}
+	dspace = H5Dget_space(dset);
+	if (dspace < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to get dataspace\n");
+		D_GOTO(out, rc);
+	}
+	vtype = H5Dget_type(dset);
+	if (vtype < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to get datatype\n");
+		D_GOTO(out, rc);
+	}
+	num_dims = H5Sget_simple_extent_dims(dspace, dims, NULL);
+	if (num_dims < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to get number of dimensions\n");
+		D_GOTO(out, rc);
+	}
+	num_attrs = dims[0];
+	D_ALLOC(attr_data, dims[0] * sizeof(struct usr_attr));
+	if (attr_data == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+	status = H5Dread(dset, vtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, attr_data);
+	if (status < 0) {
+		rc = -DER_MISC;
+		D_ERROR("failed to read attribute dataset\n");
+		D_GOTO(out, rc);
+	}
+	D_ALLOC(names, num_attrs * sizeof(char *));
+	if (names == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+	D_ALLOC(buffers, num_attrs * sizeof(void *));
+	if (buffers == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+	D_ALLOC(sizes, num_attrs * sizeof(size_t));
+	if (sizes == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(out, rc);
+	}
+	/* Set the user attribute buffers */
+	for (i = 0; i < num_attrs; i++) {
+		names[i] = attr_data[i].attr_name;
+		buffers[i] = attr_data[i].attr_val.p;
+		sizes[i] = attr_data[i].attr_val.len;
+	}
+	*_num_attrs = num_attrs;
+	*_names = names;
+	*_buffers = buffers;
+	*_sizes = sizes;
+out:
+	D_FREE(attr_data);
+	if (dset > 0)
+		H5Dclose(dset);
+	if (dspace > 0)
+		H5Sclose(dspace);
+	if (vtype > 0)
+		H5Tclose(vtype);
+	return rc;
+}
+
+
+int
+daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
```

Comment:
I think this public function needs a NULL check on filename before dereferencing